app/src/main/java/org/wikipedia/util/ThrowableUtil.kt

```diff
@@ -0,0 +1,98 @@
+            t = t.cause!!
+        }
+        return t
+    }
+
+    // TODO: replace with Apache Commons Lang ExceptionUtils.
+    private fun throwableContainsException(e: Throwable, exClass: Class<*>): Boolean {
+        var t: Throwable? = e
+        while (t != null) {
+            if (exClass.isInstance(t)) {
+                return true
+            }
+            t = t.cause
+        }
+        return false
+    }
+
+    /**
+     * DEPRECATED: This is a rarely-used function intended to sift through server error responses
+     * and pass through the relevant bits along with standardized strings in certain cases.
+     *
+     * Getting the handful of canned strings depends on processing contexts that might be null by
+     * the time we make it here.  Further, we're moving away from using raw server messages in favor
+     * of statically defined XML error views, which are safer.  This should no longer be used.
+     */
+    @JvmStatic
+    @Deprecated("")
+    fun getAppError(context: Context, e: Throwable): AppError {
+        val inner = getInnermostThrowable(e)
+        val result: AppError
+        // look at what kind of exception it is...
+        result = if (isNetworkError(e)) {
+            AppError(context.getString(R.string.error_network_error),
+                    context.getString(R.string.format_error_server_message,
+                            inner.localizedMessage))
+        } else if (e is HttpStatusException) {
+            AppError(e.message!!, e.code().toString())
+        } else if (inner is LoginFailedException || inner is CreateAccountException ||
+                inner is MwException) {
+            AppError(inner.localizedMessage!!, "")
+        } else if (throwableContainsException(e, JSONException::class.java)) {
+            AppError(context.getString(R.string.error_response_malformed),
+                    inner.localizedMessage)
+        } else {
+            // everything else has fallen through, so just treat it as an "unknown" error
+            AppError(context.getString(R.string.error_unknown),
+                    inner.localizedMessage)
+        }
+        return result
+    }
+
+    @JvmStatic
+    fun isOffline(caught: Throwable?): Boolean {
+        return (caught is UnknownHostException ||
+                caught is SocketException)
+    }
+
+    @JvmStatic
+    fun isTimeout(caught: Throwable?): Boolean {
+        return caught is SocketTimeoutException
+    }
+
+    @JvmStatic
+    fun is404(caught: Throwable): Boolean {
+        return caught is HttpStatusException && caught.code() == 404
+    }
+
+    @JvmStatic
+    fun isNetworkError(e: Throwable): Boolean {
+        return (throwableContainsException(e, UnknownHostException::class.java) ||
+                throwableContainsException(e, TimeoutException::class.java) ||
+                throwableContainsException(e, SSLException::class.java))
+    }
+
+    @Deprecated("")
+    class AppError(val error: String, val detail: String?)
```

Comment:
Considering replace/rewrite this deprecated class?

Reply:
@dbrant @sharvaniharan Do you think it is still reasonable to keep it as `deprecated`? or it is fine to display the raw server messages in certain cases and remove this `deprecated` tag?

Reply:
I agree we can just remove the deprecation..re-writing can be a fresh task and we dont need to deprecate till then.

Reply:
yep, let's remove the `deprecated` tag for now.